magnum_cluster_api/resources.py

```diff
@@ -47,6 +49,99 @@
 PLACEHOLDER = "PLACEHOLDER"
 
 
+class BaseCli:
+    APPLY_CMD = []
+    DELETE_CMD = []
+
+    def __init__(self) -> None:
+        pass
+
+    def apply(self) -> None:
+        subprocess.run(self.APPLY_CMD, capture_output=True, check=True)
+
+    def delete(self) -> None:
+        try:
+            subprocess.run(self.DELETE_CMD, capture_output=True, check=False)
+        except subprocess.CalledProcessError as e:
+            if "release: not found" in e.output:
+                return
+            raise e
+
+
+class HelmRepository(BaseCli):
+    APPLY_CMD = "helm repo add".split()
+    DELETE_CMD = "helm repo delete".split()
+    REPO_NANME = ""
+    REPO_SOURCE = ""
+
+    def __init__(self) -> None:
+        assert self.REPO_NANME
+        assert self.REPO_SOURCE
+        self.APPLY_CMD += [self.REPO_NANME, self.REPO_SOURCE]
+        self.DELETE_CMD += [self.REPO_NANME]
+
+
+class HelmRelease(BaseCli):
+    APPLY_CMD = "helm upgrade --install".split()
+    DELETE_CMD = "helm delete".split()
+    CHART_SOURCE = ""
+    RELEASE_NAME = ""
+
+    RELEASE_NAMESPACE = ""
+    EXTRA_ARGS = []
+    VALUES = {}
+
+    def __init__(self) -> None:
+        assert self.RELEASE_NAME
+        assert self.CHART_SOURCE
+        self.APPLY_CMD += [
+            self.RELEASE_NAME,
+            self.CHART_SOURCE,
+            *self.EXTRA_ARGS,
+        ]
+        self.DELETE_CMD += [self.RELEASE_NAME]
+        if self.RELEASE_NAMESPACE:
+            self.APPLY_CMD += ["-n", self.RELEASE_NAMESPACE]
+            self.DELETE_CMD += ["-n", self.RELEASE_NAMESPACE]
+        if self.VALUES:
+            self.APPLY_CMD += ["-f", self._generate_values()]
+
+    def _generate_values(self):
+        yaml_string = yaml.dump(self.VALUES)
+        values_file_path = os.path.join(
+            "/tmp", "%s-%s.yaml" % (self.RELEASE_NAME, self.RELEASE_NAMESPACE)
+        )
+        with open(values_file_path, "w") as fd:
+            fd.write(yaml_string)
+        return values_file_path
+
+
+class ClusterAutoscalerHelmRelease(HelmRelease):
+    RELEASE_NAMESPACE = "magnum-system"
+
+    def __init__(self, api, cluster) -> None:
+        cluster_name = utils.get_or_generate_cluster_api_name(api, cluster)
+        self.RELEASE_NAME = cluster_name
+
+        manifests_path = pkg_resources.resource_filename(
+            "magnum_cluster_api", "manifests"
+        )
+        self.CHART_SOURCE = os.path.join(manifests_path, "cluster-autoscaler/")
+        self.VALUES = {
```

Comment:
we should override the image as well to accomodate environments that are air-gapped.